### PR TITLE
Bugfix. Goroutine leaking

### PR DESCRIPTION
**Pull request by mpdroog**
Replaced go-routine impl with lock to simplify the code and solve the leaking.

### CODE REVIEW

**Comment by yookoala** [comment]
If I read [the specification](https://fast-cgi.github.io/spec#managing-request-ids) right, `reqeustId` is supposed to be unique among all active requests in a connection. So a `requestId` should not be reused until the previous request of the same id is inactive (releasing the id). Also the web server should attempts "to keep FastCGI request IDs small" so the application can more easily track the IDs.

In this new approach, we don't track the releasing of IDs anymore and simply loop using 1-65535 as `requestId`, am I right? While it fixes the leaking problem, it doesn't seems to do well to the specification. Although rare, 65536+ goroutine using the same id pool can potentially be using the same id of an active request. Can you convince me that this is logically impossible?

**Comment by mpdroog** [comment]
Interesting, thank you for educating me :)

> simply loop using 1-65535 as requestId, am I right?
Correct, I'm assuming when we flip to 1 again those previous requestId reservations are long gone.

> Although rare, 65536+ goroutine using the same id pool can potentially be using the same id of an active request. Can you convince me that this is logically impossible?
Hmm I see, in theory a (long) running process could still be active causing requestIds to be doubly used.

**Comment by mpdroog** [comment]
Processed your feedback and added a test to ensure a requestId isn't used twice, please check the code again.

P.S. Thank you for this awesome project

**Comment by yookoala** [comment]
The test against this is not enough. Several requirements are missed:
1. request id are supposed to start at 1 instead of 0 (as reported in #65).
2. After all ids are allocated, the pool is suppose to block `Alloc()` until a new id is released.

Especially on requirement 2, I've written a test for it. And `pool.Release()` seems to be blocked by the lock:

```go
func newIDs() *idPool {
	pool := &idPool{}
	pool.Used = make(map[uint16]struct{})
	pool.Lock = new(sync.Mutex)
	return pool
}

// If all IDs are used up, pool is supposed to block on alloc.
func TestIdPool_block(t *testing.T) {
	pool := newIDs()

	// Test allocating all ids once.
	for i := uint16(1); i < MaxUInt; i++ {
		id := pool.Alloc()
		if want, have := i, id; want != have {
			t.Errorf("expected to allocate %v, got %v", want, have)
			t.FailNow()
		}
	}

	alloc := make(chan uint16)
	go func() {
		alloc <- pool.Alloc()
	}()

	// wait some time to see if we can allocate id again
	select {
	case <-time.After(time.Millisecond * 10):
		t.Logf("alloc before release timeout as expected")
	case id := <-alloc:
		t.Errorf("allocated id unexpectedly: %v", id)
	}

	go func() {
		// release an id
		pool.Release(42)
		t.Logf("id released")
	}()

	// wait some time to see if we can allocate id again
	select {
	case <-time.After(time.Millisecond * 10):
		t.Errorf("alloc after release timeout unexpectedly")
	case id := <-alloc:
		if want, have := uint16(42), id; want != have {
			t.Errorf("expected %v, got %v", want, have)
		}
	}
}
```

File: client.go [hunk 1]
```diff
@@ -64,57 +67,46 @@ type Request struct {
...
-	// sanatize limit
-	if limit == 0 || limit > 65535 {
-		// Note: limit is the size of the pool
-		// Since 0 cannot be requestId, the effective
-		// pool is from 1 to 65535, hence size is 65535.
-		limit = 65535
+	if _, inuse := p.Used.Load(idx); inuse {
+		// Allow other go-routine to take priority
+		// to prevent spinlock here
+		runtime.Gosched()
```
Comment (yookoala) [inline client.go:91]: This context switch can solve the previous test where only 2 goroutines are in the picture. But can it solve the case where `pool.Alloc()` is called on multiple goroutines before `pool.Release()` is called? Please update the blocking test to demonstrate. Thanks.

**Review by yookoala — CHANGES_REQUESTED** [review]
Some updates are needed. Thanks for your work.

File: client.go [hunk 2]
```diff
@@ -64,57 +67,46 @@ type Request struct {
...
-	//
-	// Ref: https://fast-cgi.github.io/spec#33-records
-	ids := make(chan uint16)
-	go func(maxID uint16) {
-		for i := uint16(1); i < maxID; i++ {
-			ids <- i
-		}
-		ids <- uint16(maxID)
-	}(uint16(limit))
+	p.Used.Store(idx, struct{}{})
```
[the remark this answers is not in the document]
Comment (mpdroog) [reply client.go:95]: empty struct consumes zero bytes
https://dave.cheney.net/2014/03/25/the-empty-struct

File: client.go (hunk 1 above)
Comment (mpdroog) [reply client.go:91]: I'll have a look into extending the test.

Comment (mpdroog) [reply client.go:91]: Theoretical answer: Yes, because of the lock on Alloc.

**Comment by mpdroog** [comment]
```./client.go:72: undefined: sync.Map```
Build fails in golang 1.7+1.8 because sync.Map (1.9) is missing.

Do you want to keep supporting those older golang versions?

**Comment by yookoala** [comment]
@mpdroog: go 1.7 and 1.8 has been here for very long time. I believe it is safe to drop the support to them.

I've squash the changes here (with a few changes) to create #70. It is merged it now.
Thanks for your work.

**Pull request closed without merging**